---
title: "GLM DGE"
output:
  pdf_document: default
params:
  input_samples: "..\\data_analysis\\samples_Zuri.csv"
  regex: "Zuri"
  input_counts: "..\\data_analysis\\Zuri_oases_txi.csv"
  install: 0
  p : 0.1
  lfc : 0.1
  counts_out : "..\\data_analysis\\Zuri_oases"
  deg_out : "..\\data_analysis\\DE_lists\\Zuri_oases_"
  image_out : "..\\data_analysis\\images\\Zuri_oases\\Zuri_oases_"
  base_out: "..\\data_analysis\\"
---



```{r}
if(params$install == 1){
  if (!require("BiocManager", quietly = TRUE))
      install.packages("BiocManager")

  BiocManager::install("edgeR")
}
```

```{r}
library(edgeR)
library(pheatmap)
library(ggvenn)
library(circlize)
library(eulerr)
library(ggrepel)
library(dplyr)
```


```{r}
#check to make sure output directories exist, create them if not
deg_dir = unlist(strsplit(params$deg_out, "\\\\[^\\]*$"))[1]
image_dir = unlist(strsplit(params$image_out, "\\\\[^\\]*$"))

if(!dir.exists(deg_dir)){
  dir.create(deg_dir)
}
if(!dir.exists(image_dir)){
  dir.create(image_dir)
}
```


```{r}
#read in samples and select the ones we want
samples = read.csv(params$input_samples)
samples
```


```{r}
#for unscaled counts and length offset

#read in txi data (counts and length offset)
c_l = read.csv(params$input_counts)
#set rownames to geneids
rownames(c_l) = c_l$X
#remove column containing geneids
c_l = c_l[,-1]
#extract counts and lengths tables from dataframe
cts = c_l[,1:length(samples$Name)]
colnames(cts) = samples$Name

normMat = c_l[,(length(samples$Name)+1):(2*length(samples$Name))]
colnames(normMat) = samples$Name
#convert lengths from dataframe to matrix so tximport provided code works
normMat = as.matrix(normMat)
```



```{r}
#for unscaled counts and length offset

#follow steps from tximport vignette to prep data for edgeR

# Obtaining per-observation scaling factors for length, adjusted to avoid
# changing the magnitude of the counts.
normMat <- normMat/exp(rowMeans(log(normMat)))
normCts <- cts/normMat

# Computing effective library sizes from scaled counts, to account for
# composition biases between samples. 
eff_lib <- calcNormFactors(normCts) * colSums(normCts)

# Combining effective library sizes with the length factors, and calculating
# offsets for a log-link GLM.
normMat <- sweep(normMat, 2, eff_lib, "*")
normMat <- log(normMat)

# Creating a DGEList object for use in edgeR.
group = factor(samples$Treatment)
y <- DGEList(counts = cts, group = group)
y <- scaleOffset(y, normMat)

# filtering using the design information
keep = filterByExpr(y, group = group, min.count = 10, min.prop=0.8)
y <- y[keep, ]
```

```{r}
#remove genes flagged later in analysis as problematic
curated = read.csv("C:\\Users\\redas\\Documents\\MLML_Class_Materials\\Logan_Lab\\Thesis_data\\data_analysis\\DE_lists\\Oases_alldatatable.csv")
keep_new = which(rownames(y$counts) %in% intersect(rownames(y$counts), curated$Uniprot))
y <- y[keep_new, ]
length(rownames(y$counts))
```



```{r}
#write out filtered and normalized counts (done above)
write.csv(y$counts, paste0(params$counts_out, "_filtered_counts.csv"))
```


```{r}
#write out filtered and normalized counts (done above) as cpm to file for use in WGCNA
#See https://github.com/loganlabcsumb/SCLERA_Tutuila_Thermotolerance/blob/main/RNAseq/WGCNA/WGCNA.Rmd for details

write.csv(cpm(y, normalized.lib.size = TRUE), paste0(params$counts_out, "_filtered_cpm.csv"))
```




```{r}
#plot samples, look for outliers
points = c(15,16,17,18,19)
colors = c("green", "blue", "red", "orange", "purple")
#save output as image
png(paste0(params$image_out, "MDS_cluster.png"), width = 700)

plotMDS(y, col = colors[group], pch = points[group], labels = colnames(y))
legend("topleft", legend = levels(group), col = colors, pch = points)

dev.off()
```

```{r}
y$samples
```


```{r}
design = model.matrix(~group, data = y$samples)
design
```

```{r}
#calculate dispersion factors
y = estimateDisp(y, design, robust = TRUE)
plotBCV(y)
```


```{r}
#fit GLM
fit = glmFit(y, design, robust = TRUE)

lrt = glmLRT(fit, coef = 2:length(colnames(design)))
topTags(lrt)
summary(decideTests(lrt))
```



```{r}
#recalculate the design matrix to exclude the intercept
design = model.matrix(~0+group, data = y$samples)
colnames(design) = levels(y$samples$group)
design
```

```{r}
#calculate dispersion factors
y = estimateDisp(y, design, robust = TRUE)
plotBCV(y)
```

```{r}
#make heatmap of all samples
cpms = cpm(y, offset = y$offset, log = T)
cols = colorRampPalette(c("blue", "black", "red"))(300)

#make dataframe to indicate which samples were in which treatment for heatmap
col_anno = data.frame(treatment = samples$Treatment)
rownames(col_anno) = samples$Name

#define annotation colors for each treatment
#anno_color_opts = c(C = "#00d65c",  R = "#82b7ff", U = "#ff9289", E = "darkorange3", M = "darkorchid3")
anno_color_opts = c(Eight = "#00d65c", Four = "#82b7ff", Two = "#ff9289")

#generate vector of treatments in samples file
treatments = unique(samples$Treatment)
#get only elements of list that match current treatments so that legend does not include extraneous colors
anno_color = list(treatment = anno_color_opts[treatments])

#note that this function will not overwrite existing files to save its output. Clear out old files or write code to detect and delete them if you plan on running repeatedly
pheatmap(cpms, cluster_rows = T, scale = "row", border_color = NA, show_rownames = F, annotation_col = col_anno, main = "Total expression heatmap (z-score)", color = cols, annotation_colors = anno_color, filename = paste0(params$image_out, "total_expression_heatmap.png"))
```



```{r}
#make contrasts
contrasts = makeContrasts(TwovsEight=Two-Eight, levels = design)
```


```{r}
volcano_plot = function(FC, FDR, xlab = "logFC", ylab = "-1*log10(FDR)", main="Volcano plot", pch=20, y = F){
  if(y){
   plot(FC, -1*log10(FDR), col = ifelse(FDR<=params$p, "red", "black"), xlab = xlab, ylab = ylab, main = main, pch = 20, ylim = c(0,10)) 
  }
  else{
    plot(FC, -1*log10(FDR), col = ifelse(FDR<=params$p, "red", "black"), xlab = xlab, ylab = ylab, main = main, pch = 20)
  }
}
```


```{r}
gg_volcano_plot = function(tt, xlab = "logFC", ylab = "-1*log10(FDR)", main="Volcano plot", add_labels = TRUE){
  tt$dir = ifelse(tt$logFC<=0, "down", "up")
  #tt$size = abs(tt$logFC)*(-log10(tt$FDR))
  tt$size = abs(tt$logFC)*(-log10(tt$PValue))
  tt$gene_name = rownames(tt)
  limit = max(abs(ceiling(max(tt$logFC))), abs(floor(min(tt$logFC))))
  ret = ggplot(tt, aes(x = logFC, y = -log10(FDR))) + geom_point(aes(size = size, color = dir)) + scale_color_manual(values = c("up" = "red", "down" = "blue")) + xlim(-limit, limit) + xlab(xlab) + ylab(ylab) + ggtitle(main) + scale_size_area() + theme_bw() + theme(panel.border = element_blank(), panel.grid.minor = element_blank(), panel.grid.major = element_blank(), axis.line = element_line()) + theme(legend.position="none")
  
  if(add_labels){
    ret = ret + geom_text_repel(data = filter(tt, threshold==TRUE), aes(label = gene_name), box.padding = 0.5, max.overlaps = Inf)
  }
  
  return(ret)
}
```


```{r}
make_heatmap = function(cpms, samples_comp, sig_genes, main, color, annotation_color, filename){
  
  #subset matrix to include only significantly differentially expressed genes for the desired samples
  sig_cpm = cpms[rownames(cpms) %in% sig_genes, colnames(cpms) %in% samples_comp$Name]
  #make dataframe to indicate which samples were in which treatment for heatmap
  col_anno = data.frame(treatment = samples_comp$Treatment)
  rownames(col_anno) = samples_comp$Name
  #only keep rownames of significant genes
  #rownames(col_anno)[which(!(rownames(col_anno) %in% sig_genes))] = 
  #make heatmap
  if(length(sig_cpm) > 0){
    pheatmap(sig_cpm, scale = "row", border_color = NA, show_rownames = T, annotation_col = col_anno, main = main, color = color, annotation_colors = annotation_color, filename = filename)
  }
  
}
```



```{r}
run_test = function(fit, contrast){
  lrt = glmLRT(fit, contrast = contrasts[,contrast])
  print(topTags(lrt))
  print(summary(decideTests(lrt)))
  
  #save image as output
  png(paste0(params$image_out, contrast, "_MDplot.png"))
  
  plotMD(lrt)
  abline(h = c(-1,1), col = "blue")
  
  dev.off()

  tt = topTags(lrt, n = NULL)$table
  
  #save output as image
  png(paste0(params$image_out, contrast, "_volcano_plot.png"), width = 550)

  #volcano plots
  volcano_plot(tt$logFC, tt$FDR)

  dev.off()
  
  #save output as image
  png(paste0(params$image_out, contrast, "_volcano_plot_fdr_cutoff.png"), width = 550)

  #volcano plots
  volcano_plot(tt$logFC, tt$FDR, y = T, main = "Volcano plot, FDR > 10^-10")

  dev.off()
  
  #add column that indicates if gene is significantly differentially expressed or not
  tt$threshold = tt$FDR < params$p & abs(tt$logFC) > params$lfc
  tt$thresholdp = tt$PValue < params$p & abs(tt$logFC) > params$lfc
  
  #Fancy volcano plot
  ggsave(paste0(params$image_out, contrast, "_volcano_plot_enhanced.pdf"), plot = gg_volcano_plot(tt))
  ggsave(paste0(params$image_out, contrast, "_volcano_plot_enhanced_no_labels.pdf"), plot = gg_volcano_plot(tt, add_labels = FALSE))
  
  #write out entire toptags table
  write.csv(tt, paste0(params$deg_out, contrast, "_P", params$p, "C", params$lfc, "_full_table.csv"))
  
  #return names of genes that are significantly differentially expressed
  return(list(subset(tt, threshold == TRUE)[,-ncol(tt)], subset(tt, thresholdp == TRUE)[,-ncol(tt)]))

}
```


```{r}
write_sig_genes = function(sig_genes, comparison){
  #save significant DEGs to file
  #Upregulated genes
  write.csv(subset(sig_genes, logFC > 0), paste0(params$deg_out, comparison, "_P", params$p, "C", params$lfc, "_UP.csv"))
  #Downregulated genes
  write.csv(subset(sig_genes, logFC < 0), paste0(params$deg_out, comparison, "_P", params$p, "C", params$lfc, "_DOWN.csv"))
  #All significant genes
  write.csv(sig_genes, paste0(params$deg_out, comparison, "_P", params$p, "C", params$lfc, "_DE.csv"))
}
```

```{r}
#DEGs shared between 3 or more comparisons may cause problems here
chord = function(de_df){
  total_list = data.frame()
  data_out = data.frame()
  comps = unique(de_df$comp)
  for(i in seq_len(length(comps)-1)){
    j = i+1
    while(j <= length(comps)){
      de1 = de_df[de_df$comp == comps[i],]$genes
      de2 = de_df[de_df$comp == comps[j],]$genes
      shared = intersect(de1, de2)
      temp = data.frame(transcript_name = shared)
      row.names(temp) = shared
      write.csv(temp, paste0(params$deg_out, "P", params$p, "C", params$lfc, "_", comps[i], "_", comps[j], "_shared.csv"))
      
      data_out = rbind(data_out, data.frame(m1 = comps[i], m2 = comps[j], numShared = length(shared), percentM1 = length(shared)/length(de1), precentM2 = length(shared)/length(de2), lenM1 = length(de1), lenM2 = length(de2)))
      
      total_list = rbind(total_list, data.frame(m1 = rep(comps[i], length(shared)), m2 = rep(comps[j], length(shared)), gene_name = shared))
      j = j+1
    }
  }
  
  write.csv(total_list, paste0(params$deg_out,"P", params$p, "C", params$lfc, "_comparison_overlap.csv"))
  
  write.csv(data_out, paste0(params$base_out, params$regex, "_DEG_treatment_comparison.csv"))
  
  #make chord diagrams
  if(length(total_list) > 0){
    lengths = length(comps)
    temp = matrix(0, nrow = lengths, ncol = lengths)
    row.names(temp) = comps
    colnames(temp) = comps
    
    for(i in seq_len(nrow(data_out))){
      temp[data_out[i,1], data_out[i,2]] = data_out[i,3]
    }
    
    unmapped_1 = apply(temp, 1, sum)
    unmapped_2 = apply(temp, 2, sum)
    
    unmapped = unmapped_1 + unmapped_2
    
    length_list = unique(data.frame(m = c(data_out$m1, data_out$m2), len = c(data_out$lenM1, data_out$lenM2)))
    
    #divide by two because the unmapped genes chords start and end in the same segment, so they are essentially double-counted
    unmapped = (length_list$len - unmapped)/2
    
    names(unmapped) = comps
    
    cols = rand_color(nrow(temp)*ncol(temp))
    dim(cols) = dim(temp)
    dimnames(cols) = dimnames(temp)
    
    
    for(i in seq_len(length(unmapped))){
      temp[names(unmapped)[i], names(unmapped)[i]] = unmapped[i]
      cols[names(unmapped)[i], names(unmapped)[i]] = "#FFFFFF00"
    }
    
  png(paste0(params$image_out, "_chord_diagram.png"), width = 700, height = 700, units = "px", res = 120)
  
  chordDiagram(temp, col = cols)
  
  dev.off()
  }
}
```

```{r}
calc_deg = function(fit, test, anno_color_opts, color){
  sig_genes = run_test(fit, test)

  
  test_list = unlist(strsplit(test, "vs"))
  samples_comp = subset(samples, Treatment == test_list[1] | Treatment == test_list[2])
  
  #make heatmaps if more than 1 DEG present (otherwise doesn't work)
  if(length(rownames(sig_genes)) > 1){

    make_heatmap(cpms = cpms, samples_comp = samples_comp, sig_genes = rownames(sig_genes), main = paste0(test, " significant DEG heatmap"), color = color, annotation_color = list(treatment = anno_color_opts[test_list]), filename = paste0(params$image_out, test, "_deg_heatmap.png"))
  
    make_heatmap(cpms = cpms, samples = samples, sig_genes = rownames(sig_genes), main = paste0(test, " significant DEG heatmap"), color = color, annotation_color = list(treatment = anno_color_opts), filename = paste0(params$image_out, test, "_deg_heatmap_all_samples.png"))
  
  }
  
  write_sig_genes(sig_genes, test)
  
  return(sig_genes)
}
```

```{r}
#fit GLM
fit = glmFit(y, design, robust = TRUE)

#Test for DEG between any two groups

#define annotation colors for each treatment for heatmaps
anno_color_opts = c(C = "#00d65c",  R = "#82b7ff", U = "#ff9289")

tests = c("UvsC", "RvsC", "UvsR")

#parameters to use for subsequent tests
sig_genes = c()
sig_genes_shared = c()
sig_list = c()
sig_list_up = c()
sig_list_down = c()

for(test in tests){
  res = calc_deg(fit, test, anno_color_opts, cols)
  sig_genes = unique(sig_genes, rownames(res))
  sig_genes_shared = intersect(sig_genes, rownames(res))
  sig_list[[test]] = rownames(res)
  sig_list_up[[test]] = rownames(subset(res, logFC > 0))
  sig_list_down[[test]] = rownames(subset(res, logFC < 0))
}

make_heatmap(cpms = cpms, samples = samples, sig_genes = sig_genes, main = "Significant DEG heatmap", color = cols, annotation_color = list(treatment = anno_color_opts), filename = paste0(params$image_out, "sig_deg_heatmap_all_samples.png"))

#de_df = data.frame(comps = c(rep("UvsC", nrow(sig_genes_UC)),rep("RvsC", nrow(sig_genes_RC)),rep("UvsR", nrow(sig_genes_UR))), genes = c(rownames(sig_genes_UC),rownames(sig_genes_RC),rownames(sig_genes_UR)))

#chord(de_df)
```

```{r}
#make heatmap of significant genes shared across all treatments

#only do this if there are shared DE genes
if(length(sig_genes_shared) > 0){
  samples_comp = subset(samples, Treatment == "U" | Treatment == "R" | Treatment == "C")

  
  anno_color = list(treatment = anno_color_opts)

  make_heatmap(cpms = cpms, samples_comp = samples_comp, sig_genes = sig_genes_shared, main = "shared DEG heatmap", color = cols, annotation_color = anno_color, filename = paste0(params$image_out, "shared_deg_heatmap.png"))

}
```


```{r}
#Venn diagrams of all DEGs

png(paste0(params$image_out, "treatment_venn_all.png"))

plot(euler(sig_list, shape = "ellipse"), quantities = TRUE)

dev.off()

png(paste0(params$image_out, "treatment_venn_up.png"))

plot(euler(sig_list_up, shape = "ellipse"), quantities = TRUE)

dev.off()

png(paste0(params$image_out, "treatment_venn_down.png"))

plot(euler(sig_list_down, shape = "ellipse"), quantities = TRUE)

dev.off()
```















































































































