---
title: "R Notebook"
output:
  pdf_document: default
  html_notebook: default
---



```{r}

##Tximport driver chunk

#input variable names for use in script

#types:
#samples: all samples
#no_batch: all samples except those with batch effects
#batch: all C, U, and R samples, including those with batch effects
#red: Reduced. Only C, U, and R samples that don't show batch effects

types = c()

#types = append(types, "all")
#types = append(types, "no_batch")
#types = append(types, "batch")
types = append(types, "red")

#Regex of samples names we want to keep
tissues = c()

tissues = append(tissues, "CB")
tissues = append(tissues, "CG")
tissues = append(tissues, "CM")
tissues = append(tissues, "CL")

#genome = TRUE
genome = FALSE

#install tximport?
install = 0



for(regex in tissues){
  for(type in types){
    if(genome == TRUE){
      #file to write csv to
      out = paste0("..\\data_analysis\\genome_", regex, "_", type, "_txi.csv")
      
      samples_location = "..\\mapping\\genome"
      
      #location of gene_trans_map
      gtm = "..\\annotation\\genome_gene_trans_map.csv"
      
      #ignore . in transcript name
      txversion = FALSE
    }else{
      #file to write csv to
      out = paste0("..\\data_analysis\\", regex, "_", type, "_txi.csv")
      
      samples_location = "..\\mapping"
      
      #location of gene_trans_map
      gtm = "..\\annotation\\gene_trans_map.csv"
      
      #ignore . in transcript name
      txversion = TRUE
    }
    
    #names of file with salmon output folders
    input_samples = input_samples = paste0("..\\data_analysis\\samples_", type, ".csv")

    rmarkdown::render(input = "tximport.Rmd", params = list(out = out, gtm = gtm, samples_location = samples_location, input_samples = input_samples, regex = regex, txversion = txversion))

    install = 0
    
  }
}

```


```{r}
##DGE GLM driver chunk
#add heatmap and pearson correlation
#get DEG lists -> send to Holly
#do UvsE and fluctuating vs all 3 static comparisons in no_batch

#may need to check box for Tools -> Global options -> Sweave -> LaTeX editing and compilation -> Use tinytex when compiling .tex files for this to work


#set parameters for compilation

#input data for run


#types:
#samples: all samples
#no_batch: all samples except those with batch effects
#batch: all C, U, and R samples, including those with batch effects
#red: Reduced. Only C, U, and R samples that don't show batch effects

types = c()

#types = append(types, "all")
#types = append(types, "no_batch")
#types = append(types, "batch")
types = append(types, "red")
#types = append(types, "corrected")

#Regex of samples names we want to keep
tissues = c()

tissues = append(tissues, "CB")
tissues = append(tissues, "CG")
tissues = append(tissues, "CM")
tissues = append(tissues, "CL")

genome = "genome_"
#genome = ""

#if we want to install the package. Set to 1 to install
install = 0

for(regex in tissues){
  for(type in types){
    #samples description file
    input_samples = paste0("..\\data_analysis\\samples_", type, ".csv")

    input_counts = paste0("..\\data_analysis\\", genome, regex, "_", type, "_txi.csv")
    
    #Path to output directory for filtered gene lists for WGCNA
    counts_out = paste0("..\\data_analysis\\", genome, regex, "_", type, "_filtered_counts.csv")
    
    #Path to output directory for DEG lists
    deg_out = paste0("..\\data_analysis\\DE_lists\\", genome, regex, "_", type, "_")
    
    #Path to output directory for images
    image_out = paste0("..\\data_analysis\\images\\", genome, regex, "_", type, "\\", genome, regex, "_", type, "_")

    #path to and name of output report file
    #if it errors, try using just the file name output and moving it where you want afterwards. I was having some issues getting this to work when it had to write the pdf to a different directory than the .Rmd file, although it wrote the .tex file to the location just fine
    report_output = paste0(genome, regex, "_", type, "_LRT.pdf")

    #render script in new R session to isolate it
    xfun::Rscript_call(rmarkdown::render, list(input = "DGE_glm.Rmd", params = list(input_samples = input_samples, regex = regex, input_counts = input_counts, install = install, counts_out = counts_out, deg_out = deg_out, image_out = image_out), output_file = report_output, output_format = 'pdf_document'))
  
    #we only want to install on the first iteration, so set install to 0 now
    install = 0

  }
}

```

```{r}
##DGE within treatment driver chunk

#may need to check box for Tools -> Global options -> Sweave -> LaTeX editing and compilation -> Use tinytex when compiling .tex files for this to work


#set parameters for compilation

#input data for run


#types:
#samples: all samples
#no_batch: all samples except those with batch effects
#batch: all C, U, and R samples, including those with batch effects
#red: Reduced. Only C, U, and R samples that don't show batch effects

types = c()

#types = append(types, "all")
#types = append(types, "no_batch")
#types = append(types, "batch")
types = append(types, "red")

#Regex of samples names we want to keep
tissues = c()

tissues = append(tissues, "CB")
tissues = append(tissues, "CG")
tissues = append(tissues, "CM")
tissues = append(tissues, "CL")

#Upregulated, downregulated, and all lists
directions = c()

directions = append(directions, "UP")
directions = append(directions, "DOWN")
directions = append(directions, "DE")

#treatments that were performed
treatments = c()

treatments = append(treatments, "UvsC")
treatments = append(treatments, "RvsC")
treatments = append(treatments, "UvsR")

#P-value cutoff of DEG list
p = 0.05

#log-fold change cutoff of DEG list
lfc = 1

#genome = ""
genome = "genome_"

for(treatment in treatments){
  for(direction in directions){
    for(type in types){
      input_degs = c()
      for(regex in tissues){
        input_degs = append(input_degs, paste0("..\\data_analysis\\DE_lists\\",genome, regex, "_", type, "_", treatment, "_P", p, "C", lfc, "_", direction, ".csv"))
      }
      image_out = paste0("..\\data_analysis\\images\\", treatment, "\\", genome, treatment, "_", type, "_", direction, "_")
      #render script in new R session to isolate it
    xfun::Rscript_call(rmarkdown::render, list(input = "within_treatment_analysis.Rmd", params = list(labels = tissues, input_degs = input_degs, image_out = image_out)))
    }
  }
}
```

```{r}
#DGE GO analysis chunk

#set parameters for compilation

#input data for run


#types:
#samples: all samples
#no_batch: all samples except those with batch effects
#batch: all C, U, and R samples, including those with batch effects
#red: Reduced. Only C, U, and R samples that don't show batch effects

types = c()

#types = append(types, "all")
#types = append(types, "no_batch")
#types = append(types, "batch")
types = append(types, "red")

#Regex of samples names we want to keep
tissues = c()

tissues = append(tissues, "CB")
tissues = append(tissues, "CG")
tissues = append(tissues, "CM")
tissues = append(tissues, "CL")

#genome = "genome_"
genome = ""

p = 0.05
c = 1

#treatment contrasts to analyze
comparisons = c()

comparisons = append(comparisons, "UvsC")
comparisons = append(comparisons, "RvsC")
comparisons = append(comparisons, "UvsR")

directions = c()

directions = append(directions, "DE")
directions = append(directions, "UP")
directions = append(directions, "DOWN")

for(type in types){
  for(tissue in tissues){
    for(comparison in comparisons){
      for(direction in directions){
       de = paste0("..\\data_analysis\\DE_lists\\", genome, tissue, "_", type, "_", comparison, "_P", p, "C", c, "_", direction, ".csv") 

       counts = paste0("..\\data_analysis\\", genome, tissue, "_", type, "_txi.csv")
       
       baseout = paste0("..\\data_analysis\\images\\", genome, tissue, "_", type, "\\", genome, tissue, "_", type, "_", comparison, "_", direction, "_")
       
       listout = paste0("..\\data_analysis\\GO_lists\\", genome, tissue, "_", type, "_", comparison, "_P", p, "C", c, "_", direction, "_")
       
       gtm = paste0("..\\annotation\\", genome, "gene_trans_map.csv")
       
       info = paste0("..\\annotation\\sw_", genome, "gene_info.csv")
       
       xfun::Rscript_call(rmarkdown::render, list(input = "GO_analysis.Rmd", params = list(de = de, counts = counts, baseout = baseout, listout = listout, gtm = gtm, info = info)))
      }
    }
  }
}


```


```{r}
#DGE KEGG analysis chunk

#set parameters for compilation

#input data for run


#types:
#samples: all samples
#no_batch: all samples except those with batch effects
#batch: all C, U, and R samples, including those with batch effects
#red: Reduced. Only C, U, and R samples that don't show batch effects

types = c()

#types = append(types, "all")
#types = append(types, "no_batch")
#types = append(types, "batch")
types = append(types, "red")

#Regex of samples names we want to keep
tissues = c()

tissues = append(tissues, "CB")
tissues = append(tissues, "CG")
tissues = append(tissues, "CM")
tissues = append(tissues, "CL")

genome = "genome_"
#genome = ""

p = 0.05
c = 1

#treatment contrasts to analyze
comparisons = c()

#comparisons = append(comparisons, "UvsC")
#comparisons = append(comparisons, "RvsC")
comparisons = append(comparisons, "UvsR")

directions = c()

directions = append(directions, "DE")
directions = append(directions, "UP")
directions = append(directions, "DOWN")

for(type in types){
  for(tissue in tissues){
    for(comparison in comparisons){
      for(direction in directions){
       de = paste0("..\\data_analysis\\DE_lists\\", genome, tissue, "_", type, "_", comparison, "_P", p, "C", c, "_", direction, ".csv") 

       counts = paste0("..\\data_analysis\\", genome, tissue, "_", type, "_txi.csv")
       
       baseout = paste0("..\\data_analysis\\images\\", genome, tissue, "_", type, "\\", genome, tissue, "_", type, "_", comparison, "_", direction, "_")
       
       listout = paste0("..\\data_analysis\\GO_lists\\", genome, tissue, "_", type, "_", comparison, "_P", p, "C", c, "_", direction, "_")
       
       gtm = paste0("..\\annotation\\", genome, "gene_trans_map.csv")
       
       info = paste0("..\\annotation\\sw_", genome, "kegg_info.csv")
       
       xfun::Rscript_call(rmarkdown::render, list(input = "GO_analysis.Rmd", params = list(de = de, counts = counts, baseout = baseout, listout = listout, gtm = gtm, info = info, kegg = 1)))
      }
    }
  }
}


```


```{r}
#WGCNA driver chunk


#may need to check box for Tools -> Global options -> Sweave -> LaTeX editing and compilation -> Use tinytex when compiling .tex files for this to work


#set parameters for compilation

#input data for run


#types:
#samples: all samples
#no_batch: all samples except those with batch effects
#batch: all C, U, and R samples, including those with batch effects
#red: Reduced. Only C, U, and R samples that don't show batch effects

types = c()

#types = append(types, "all")
#types = append(types, "no_batch")
#types = append(types, "batch")
types = append(types, "red")
#types = append(types, "corrected")

#Regex of samples names we want to keep
tissues = c()

tissues = append(tissues, "CB")
#tissues = append(tissues, "CG")
#tissues = append(tissues, "CM")
#tissues = append(tissues, "CL")

#genome = "genome_"
genome = ""

sw = paste0("..\\annotation\\sw_", genome, "gene_info.csv") 

for(regex in tissues){
  for(type in types){
    
    input_counts = paste0("..\\data_analysis\\", genome, regex, "_", type, "_filtered_counts.csv")
    
    input_samples = paste0("..\\data_analysis\\samples_", type, ".csv")
    
    traits = paste0("..\\data_analysis\\", regex, "_traits.csv")
    

    #render script in new R session to isolate it
    xfun::Rscript_call(rmarkdown::render, list(input = "WGCNA_analysis.Rmd", params = list(input_samples = input_samples, regex = regex, genome = genome, input_counts = input_counts, traits = traits, sw = sw)))

  }
}

```



```{r}
#WGCNA GO analysis chunk

#set parameters for compilation

#input data for run


#types:
#samples: all samples
#no_batch: all samples except those with batch effects
#batch: all C, U, and R samples, including those with batch effects
#red: Reduced. Only C, U, and R samples that don't show batch effects

types = c()

#types = append(types, "all")
#types = append(types, "no_batch")
#types = append(types, "batch")
types = append(types, "red")

#Regex of samples names we want to keep
tissues = c()

tissues = append(tissues, "CB")
tissues = append(tissues, "CG")
tissues = append(tissues, "CM")
tissues = append(tissues, "CL")

genome = ""
genome_inc = 0
#genome = "genome_"
#genome_inc = 1

for(type in types){
  for(tissue in tissues){
    sig_modules = list.files("..\\data_analysis\\WGCNA_output\\", pattern = paste0("^", genome, tissue, "_sig"), full.names = TRUE)
    for(sig_module in sig_modules){

       counts = paste0("..\\data_analysis\\", genome, tissue, "_", type, "_txi.csv")
       if(strsplit(sig_module, "_")[[1]][7+genome_inc] == "top"){
         baseout = paste0("..\\data_analysis\\images\\WGCNA\\", genome, tissue, "_", type, "\\", genome, tissue, "_", type, "_", strsplit(sig_module, "_")[[1]][6+genome_inc], "_top_")
         
         listout = paste0("..\\data_analysis\\GO_lists\\WGCNA\\", genome, tissue, "_", type, "_", strsplit(sig_module, "_")[[1]][6+genome_inc], "_top_")
       }else{
          baseout = paste0("..\\data_analysis\\images\\WGCNA\\", genome, tissue, "_", type, "\\", genome, tissue, "_", type, "_", strsplit(sig_module, "_")[[1]][6+genome_inc], "_")
          
          listout = paste0("..\\data_analysis\\GO_lists\\WGCNA\\", genome, tissue, "_", type, "_", strsplit(sig_module, "_")[[1]][6+genome_inc], "_")
          
       }
       
       gtm = paste0("..\\annotation\\", genome, "gene_trans_map.csv")
       
       info = paste0("..\\annotation\\sw_", genome, "gene_info.csv")
       
       xfun::Rscript_call(rmarkdown::render, list(input = "GO_analysis.Rmd", params = list(de = sig_module, counts = counts, baseout = baseout, listout = listout, gtm = gtm, info = info)))
    }
  }
}
```

```{r}
#WGCNA KEGG analysis chunk

#set parameters for compilation

#input data for run


#types:
#samples: all samples
#no_batch: all samples except those with batch effects
#batch: all C, U, and R samples, including those with batch effects
#red: Reduced. Only C, U, and R samples that don't show batch effects

types = c()

#types = append(types, "all")
#types = append(types, "no_batch")
#types = append(types, "batch")
types = append(types, "red")

#Regex of samples names we want to keep
tissues = c()

tissues = append(tissues, "CB")
tissues = append(tissues, "CG")
tissues = append(tissues, "CM")
tissues = append(tissues, "CL")

genome = ""
genome_inc = 0
#genome = "genome_"
#genome_inc = 1

for(type in types){
  for(tissue in tissues){
    sig_modules = list.files("..\\data_analysis\\WGCNA_output\\", pattern = paste0("^", genome, tissue, "_sig"), full.names = TRUE)
    for(sig_module in sig_modules){

       counts = paste0("..\\data_analysis\\", genome, tissue, "_", type, "_txi.csv")
       if(strsplit(sig_module, "_")[[1]][7+genome_inc] == "top"){
         baseout = paste0("..\\data_analysis\\images\\WGCNA\\", genome, tissue, "_", type, "\\", genome, tissue, "_", type, "_", strsplit(sig_module, "_")[[1]][6+genome_inc], "_top_")
         
         listout = paste0("..\\data_analysis\\GO_lists\\WGCNA\\", genome, tissue, "_", type, "_", strsplit(sig_module, "_")[[1]][6+genome_inc], "_top_")
       }else{
          baseout = paste0("..\\data_analysis\\images\\WGCNA\\", genome, tissue, "_", type, "\\", genome, tissue, "_", type, "_", strsplit(sig_module, "_")[[1]][6+genome_inc], "_")
          
          listout = paste0("..\\data_analysis\\GO_lists\\WGCNA\\", genome, tissue, "_", type, "_", strsplit(sig_module, "_")[[1]][6+genome_inc], "_")
          
       }
       
       gtm = paste0("..\\annotation\\", genome, "gene_trans_map.csv")
       
       info = paste0("..\\annotation\\sw_", genome, "kegg_info.csv")
       
       xfun::Rscript_call(rmarkdown::render, list(input = "GO_analysis.Rmd", params = list(de = sig_module, counts = counts, info = info, baseout = baseout, listout = listout, kegg = 1, gtm = gtm)))
    }
  }
}
```


```{r}
##NMDS driver chunk

#may need to check box for Tools -> Global options -> Sweave -> LaTeX editing and compilation -> Use tinytex when compiling .tex files for this to work


#set parameters for compilation

#input data for run


#types:
#samples: all samples
#no_batch: all samples except those with batch effects
#batch: all C, U, and R samples, including those with batch effects
#red: Reduced. Only C, U, and R samples that don't show batch effects

types = c()

#types = append(types, "all")
#types = append(types, "no_batch")
#types = append(types, "batch")
types = append(types, "red")
#types = append(types, "corrected")

#Regex of samples names we want to keep
tissues = c()

tissues = append(tissues, "CB")
tissues = append(tissues, "CG")
tissues = append(tissues, "CM")
tissues = append(tissues, "CL")

genome = "genome_"
#genome = ""

for(regex in tissues){
  for(type in types){
    counts = paste0("..\\data_analysis\\", genome, regex, "_", type, "_filtered_counts.csv")
    
    samples = paste0("..\\data_analysis\\samples_", type, ".csv")
    
    image_out = paste0("..\\data_analysis\\images\\", genome, regex, "_", type, "\\", genome, regex, "_", type)

    #render script in new R session to isolate it
    xfun::Rscript_call(rmarkdown::render, list(input = "nmds.Rmd", params = list(counts = counts, samples = samples, regex = regex, image_out = image_out)))

  }
}

```


```{r}
##interindividual variation driver chunk

#set parameters for compilation

#input data for run


#types:
#samples: all samples
#no_batch: all samples except those with batch effects
#batch: all C, U, and R samples, including those with batch effects
#red: Reduced. Only C, U, and R samples that don't show batch effects

types = c()

#types = append(types, "all")
#types = append(types, "no_batch")
#types = append(types, "batch")
types = append(types, "red")

#Regex of samples names we want to keep
tissues = c()

tissues = append(tissues, "CB")
tissues = append(tissues, "CG")
tissues = append(tissues, "CM")
tissues = append(tissues, "CL")

genome = "genome_"
#genome = ""

kegg_info = paste0("..\\annotation\\sw_", genome, "kegg_info.csv")


for(regex in tissues){
  for(type in types){
    
    counts = paste0("..\\data_analysis\\",genome, regex, "_", type, "_", "filtered_counts.csv")
    
    samples = paste0("..\\data_analysis\\samples_", type, ".csv")
    
    base_out = paste0("..\\data_analysis\\MAD\\", genome, regex, "_", type, "_")

    #render script in new R session to isolate it
    xfun::Rscript_call(rmarkdown::render, list(input = "inter_individual_variation.Rmd", params = list(kegg_info = kegg_info, counts = counts, regex = regex, samples = samples, base_out = base_out)))

  }
}
```



```{r}
##dbRDA driver chunk

#may need to check box for Tools -> Global options -> Sweave -> LaTeX editing and compilation -> Use tinytex when compiling .tex files for this to work


#set parameters for compilation

#input data for run


#types:
#samples: all samples
#no_batch: all samples except those with batch effects
#batch: all C, U, and R samples, including those with batch effects
#red: Reduced. Only C, U, and R samples that don't show batch effects

types = c()

#types = append(types, "all")
#types = append(types, "no_batch")
#types = append(types, "batch")
types = append(types, "red")
#types = append(types, "corrected")

#Regex of samples names we want to keep
tissues = c()

tissues = append(tissues, "CB")
tissues = append(tissues, "CG") # sample size is too small for this tissue in red dataset, so exclude. Otherwise would be overparameterized (# of traits > # samples), so no residuals and can't do statistical analysis
tissues = append(tissues, "CM")
tissues = append(tissues, "CL")

#genome = "genome_"
genome = ""

for(regex in tissues){
  for(type in types){
    counts = paste0("..\\data_analysis\\", genome, regex, "_", type, "_filtered_counts.csv")
    
    traits = paste0("..\\data_analysis\\", regex, "_traits.csv")
    
    input_samples = paste0("..\\data_analysis\\samples_", type, ".csv")
    
    outfile = paste0("..\\data_analysis\\dbRDA\\", genome, regex, "_", type)

    #render script in new R session to isolate it
    xfun::Rscript_call(rmarkdown::render, list(input = "dbRDA.Rmd", params = list(counts = counts, traits = traits, input_samples = input_samples, regex = regex, outfile = outfile)))

  }
}
```




```{r}

##DGE Pairwise driver chunk

#may need to check box for Tools -> Global options -> Sweave -> LaTeX editing and compilation -> Use tinytex when compiling .tex files for this to work


#set parameters for compilation

#input data for run


#types:
#samples: all samples
#no_batch: all samples except those with batch effects
#batch: all C, U, and R samples, including those with batch effects
#red: Reduced. Only C, U, and R samples that don't show batch effects

types = c()

#types = append(types, "all")
#types = append(types, "no_batch")
#types = append(types, "batch")
types = append(types, "red")

#Regex of samples names we want to keep
tissues = c()

tissues = append(tissues, "CB")
tissues = append(tissues, "CG")
tissues = append(tissues, "CM")
tissues = append(tissues, "CL")

for(regex in tissues){
  for(type in types){
    #samples description file
    input_samples = paste0("..\\data_analysis\\samples_", type, ".csv")

    input_counts = paste0("..\\data_analysis\\", regex, "_", type, "_txi.csv")

    #if we want to install the package. Set to 1 to install
    install = 0

    #path to and name of output file
    #if it errors, try using just the file name output and moving it where you want afterwards. I was having some         issues getting this to work when it had to write the pdf to a different directory than the .Rmd file, although it     wrote the .tex file to the location just fine
    output = paste0(regex, "_", type, "_pairwise", ".pdf")

    rmarkdown::render(input = "DGE_pairwise.Rmd", params = list(input_samples = input_samples, regex = regex, input_counts = input_counts, install = install), output_file = output, output_format = 'pdf_document')
    
    #we only want to install on the first iteration, so set install to 0 now
    install = 0
  
  }
}

```




































































































