---
title: "WGCNA"
output: 
  pdf_document: default
params:
  input_counts: "..\\data_analysis\\genome_CM_red_filtered_counts.csv"
  input_samples: "..\\data_analysis\\samples_red.csv"
  regex: "CM"
  genome: "genome_"
  traits: "..\\data_analysis\\CM_traits.csv"
  sw: "..\\annotation\\sw_genome_gene_info.csv"
---


```{r}
library(WGCNA)
options(stringsAsFactors = FALSE)
library(ggplot2)
library(egg)
library(tidyverse)
```

```{r}
#read in samples and select the ones we want
samples = read.csv(params$input_samples)
samples = samples[grep(params$regex, samples$Name), ]
samples
```

```{r}
data = read.csv(params$input_counts)
#set rownames to geneids
rownames(data) = data$X
#remove column containing geneids
data =  data[,-1]
data
```

```{r}
#At the end of this step, we want genes as columns and rows as samples. Rows should only contain expression data, sample names should be rownames
dat0 = as.data.frame(t(data))
names(dat0) = rownames(data)
rownames(dat0) = names(data)
dat0
```

```{r}
gsg = goodSamplesGenes(dat0, verbose = 3)
gsg$allOK
```

```{r}
if(!gsg$allOK){
  if(sum(!gsg$goodGenes) > 0)
    printFlush(paste("Removing genes:", paste(names(dat0)[!gsg$goodGenes], collapse = ", ")))
  if(sum(!gsg$goodSamples>0))
    printFlush(paste("Removing samples:", paste(rownames(dat0)[!gsg$goodSamples], collapse = ", ")))
  dat0 = dat0[gsg$goodSamples, gsg$goodGenes]
}
```

```{r}
sampleTree = hclust(dist(dat0), method = "average")
par(cex = 0.6)
par(mar = c(0,4,2,0))

png(paste0("..\\data_analysis\\images\\WGCNA\\", params$genome, params$regex, "_sample_clustering_dendrogram.png"), width = 700)

plot(sampleTree, main = "Sample clustering to detect outliers", sub = "", xlab = "", cex.lab = 1.5, cex.axis = 1.5, cex.main = 2)

dev.off()
```

```{r}
if(params$genome != ""){
  if(params$regex == "CG" || params$regex == "CL"){
    datExpr = dat0
    nGenes = ncol(datExpr)
    nSamples = nrow(datExpr)
    datExpr
  }else {
    if(params$regex == "CB"){
      clust = cutreeStatic(sampleTree, cutHeight = 40000, minSize = 5)
    }
    if(params$regex == "CM"){
      clust = cutreeStatic(sampleTree, cutHeight = 13000, minSize = 5)
    }
    table(clust)
    #clust 1 contains the samples we want to keep
    keepSamples = (clust == 1)
    datExpr = dat0[keepSamples, ]
    nGenes = ncol(datExpr)
    nSamples = nrow(datExpr)
    datExpr
    }
  }else{
  if(params$regex == "CL"){
    datExpr = dat0
    nGenes = ncol(datExpr)
    nSamples = nrow(datExpr)
    datExpr
  }else {
    if(params$regex == "CG"){
      clust = cutreeStatic(sampleTree, cutHeight = 15000, minSize = 5)
    }else if(params$regex == "CB"){
      clust = cutreeStatic(sampleTree, cutHeight = 30000, minSize = 5)
    }else if(params$regex == "CM"){
      clust = cutreeStatic(sampleTree, cutHeight = 11000, minSize = 5)
    }
    table(clust)
    #clust 1 contains the samples we want to keep
    keepSamples = (clust == 1)
    datExpr = dat0[keepSamples, ]
    nGenes = ncol(datExpr)
    nSamples = nrow(datExpr)
    datExpr
  }
}



```


```{r}
traits = read.csv(params$traits)
dim(traits)
names(traits)

#get rid of unneeded columns
#allTraits = traits[, -c(31,16)]
#allTraits = allTraits[, c(2, 11:36)]
allTraits = traits
dim(allTraits)
names(allTraits)

fs = rownames(datExpr)
#make sure keys match between tables (same sample IDs used in Phys and gene expression tables)
traitRows = match(fs, allTraits$Name)
datTraits = allTraits[traitRows, -1]
rownames(datTraits) = allTraits[traitRows, 1]

collectGarbage()
datTraits
```

```{r}
#visualization of clinical traits relation to sample dendrogram

sampleTree2 = hclust(dist(datExpr), method = "average")
traitColors = numbers2colors(datTraits, signed = FALSE)
plotDendroAndColors(sampleTree2, traitColors, groupLabels = names(datTraits), main = "Sample dendrogram and heatmap")
```


#network construction and module detection

```{r}
#choose set of soft-thresholding powers
powers = c(c(1:10), seq(from = 12, to = 20, by = 2))

#call the network topology analysis function
sft = pickSoftThreshold(datExpr, powerVector = powers, verbose = 5)

#sizeGrWindow(9,5)
par(mfrow = c(1,2))
cex1 = 0.9

#scale-free topology fit index as a function of the soft-thresholding power
png(paste0("..\\data_analysis\\images\\WGCNA\\", params$genome, params$regex, "_soft_thresholding.png"))

plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2], xlab = "Soft Threshold (power)", ylab = "Scale Free Topology Model Fit,signed R^2", type = "n", main = paste("Scale independence"))

text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2], labels = powers, cex = cex1, col = "red")

#this line corresponds to using an R^2 cut-off of h
abline(h=0.90, col= "red")

dev.off()

#mean connectivity as a function of the soft-thresholding power
png(paste0("..\\data_analysis\\images\\WGCNA\\", params$genome, params$regex, "_mean_connectivity.png"))

plot(sft$fitIndices[5:20,1], sft$fitIndices[5:20,5], xlab = "Soft Threshold (power)", ylab = "Mean connectivity", type = "n", main = paste("Mean connectivity"))
text(sft$fitIndices[,1], sft$fitIndices[,5], labels = powers, cex = cex1, col = "red")

dev.off()
```

```{r}
#use 9 if <20 samples, scale free topology index fails to reach 0.8 at <15, and mean connectivity remains in the 100s. See WGCNA FAQ's for more info
if(params$genome != ""){
  if(params$regex == "CB"){
    spower = 7
  }else if(params$regex == "CG"){
    spower = 9
  }else if(params$regex == "CL"){
    spower = 12
  }else if(params$regex == "CM"){
    spower = 9
  }
}else{
  if(params$regex == "CB"){
    spower = 8
  }else if(params$regex == "CG"){
    spower = 9
  }else if(params$regex == "CL"){
    spower = 12
  }else if(params$regex == "CM"){
    spower = 9
  } 
}

```


```{r}
#construct gene networks and identify modules
#pick power = 10 because it seems to even off there
net = blockwiseModules(datExpr, power = spower, TOMType = "unsigned", minModuleSize = 30, reassignThreshold = 0, mergeCutHeight = 0.25, numericLabels = TRUE, pamRespectsDendro = FALSE, verbose = 3)
```

```{r}
#how many modules there are and how many genes they contain in descending order. Module 0 is the genes that didn't fit into any of the other modules (the leftover ones)
table(net$colors)
```

```{r}
sizeGrWindow(12,9)
mergedColors = labels2colors(net$colors)
plotDendroAndColors(net$dendrograms[[1]], mergedColors[net$blockGenes[[1]]], "Module colors", dendroLabels = FALSE, hang = 0.03, addGuide = TRUE, guideHang = 0.05)
```

```{r}
moduleLabels = net$colors
moduleColors = labels2colors(net$colors)
MEs = net$MEs
geneTree = net$dendrograms[[1]]
```

#Relating modules to external info and identifying important genes

```{r}
#Recalculate MEs with color labels
MEs = moduleEigengenes(datExpr, moduleColors)$eigengenes
MEs = orderMEs(MEs)
moduleTraitCor = cor(MEs, datTraits, use = "p")
moduleTraitPvalue = corPvalueStudent(moduleTraitCor, nSamples)
```

```{r}
write.csv(moduleTraitPvalue, paste0("..\\data_analysis\\", "WGCNA_output\\", params$genome, params$regex, "_module_trait_info.csv"))

write.csv(moduleTraitCor, paste0("..\\data_analysis\\", "WGCNA_output\\", params$genome, params$regex, "_module_trait_cor.csv"))
```


```{r}
#graphical representation
#display correlation and p value if <= 0.05, - if not
textMatrix = paste(ifelse(moduleTraitPvalue <= 0.05, paste(signif(moduleTraitCor, 2), "\n(", signif(moduleTraitPvalue, 1), ")",  sep = ""), "-"))
dim(textMatrix) = dim(moduleTraitCor)

#display correlation values within a heatmap
png(paste0("..\\data_analysis\\images\\WGCNA\\", params$genome, params$regex, "_labeled_heatmap.png"), width = 1000, height = 700)

par(mar = c(9, 12.5, 3, 3))
labeledHeatmap(Matrix = moduleTraitCor, xLabels = names(datTraits), yLabels = names(MEs), ySymbols = names(MEs), colorLabels = FALSE, colors = greenWhiteRed(50), textMatrix = textMatrix, setStdMargins = FALSE, cex.text = 0.5, zlim = c(-1,1), main = paste("Module-trait relationships"))

dev.off()
```

```{r}
convert_to_index = function(i, dim1){
  return(c(((i-1)%%dim1)+1, ceiling(i/dim1)))
}
```


```{r}
#convert raw indices of significant associations between modules and traits into row, column pairs (row = module, column  = trait) that can be indexed into matrix
sig_pos = lapply(which(textMatrix != "-"), convert_to_index, dim(textMatrix)[1])

#select significant associations with categorical variables
sig_categorical = compact(mapply(sig_pos, FUN = function(x){if(x[2] <= 3){return(x)}}))

#select significant associations with continuous variables
sig_continuous = compact(mapply(sig_pos, FUN = function(x){if(x[2] > 3){return(x)}}))
```


```{r}
weight = datTraits
#names(weight) = "weight"
#names (colors) of the modules
modNames = substring(names(MEs), 3)

geneModuleMembership = as.data.frame(cor(datExpr, MEs, use = "p"))
MMPvalue = as.data.frame(corPvalueStudent(as.matrix(geneModuleMembership), nSamples))

names(geneModuleMembership) = paste("MM", modNames, sep = "")
names(MMPvalue) = paste("p.MM", modNames, sep = "")

geneTraitSignificance = as.data.frame(cor(datExpr, weight, use = "p"))
GSPvalue = as.data.frame(corPvalueStudent(as.matrix(geneTraitSignificance), nSamples))

names(geneTraitSignificance) = paste("GS.", names(weight), sep = "")
names(GSPvalue) = paste("p.GS", names(weight), sep = "")

```

```{r}
#Intramodular analysis: identifying genes with high GS and MM

module = "pink"
trait = 2
traitName = unlist(strsplit(names(geneTraitSignificance[trait]), ".", fixed = TRUE))[trait]
column = match(module, modNames)
moduleGenes = moduleColors == module

sizeGrWindow(7,7)
par(mfrow = c(1,1))
verboseScatterplot(abs(geneModuleMembership[moduleGenes, column]), abs(geneTraitSignificance[moduleGenes, trait]), xlab = paste("Module Membership in", module, "module"), ylab = paste0("Gene significance for ", traitName), main = paste("Module membership vs gene significance\n"), cex.main = 1.2, cex.lab = 1.2, cex.axis = 1.2, col = module)

```

```{r}
sw = read.csv (params$sw)
annot = sw [, -1]
```

```{r}
#summary output of network analysis results
dim(annot)
names(annot)
probes = names(datExpr)
probes2annot = match(probes, annot$accession_id)
#number of probes without annotation
sum(is.na(probes2annot))
#if all have annotation, should return 0
```

```{r}
geneInfo = data.frame(transcript_name = probes, geneSymbol = annot$accession_id[probes2annot], moduleColor = moduleColors, geneTraitSignificance, GSPvalue)
#order modules by significance for weight
modOrder = order(-abs(cor(MEs, weight[,trait], use = "p")))
#add module membership info in the chosen order
for(mod in 1:ncol(geneModuleMembership)){
  oldNames = names(geneInfo)
  geneInfo = data.frame(geneInfo, geneModuleMembership[, modOrder[mod]], MMPvalue[, modOrder[mod]])
  names(geneInfo) = c(oldNames, paste("MM.", modNames[modOrder[mod]], sep = ""), paste("p.MM", modNames[modOrder[mod]], sep = ""))
}

#order the genes in geneInfo first by module color, then by geneTraitSignificance
geneOrder = order(geneInfo$moduleColor, -abs(geneInfo[,trait+3]))
geneInfo = geneInfo[geneOrder, ]
```

```{r}
#select the vector index of significant cells in the Pvalue matrix, then mod by the number of modules (rows) to convert to row index. Then remove duplicates in case a module was significantly associated with more than one trait
sigRows = unique((which(moduleTraitPvalue < 0.05)-1)%%length(table(net$colors))+1)

#select only those modules that were signicantly correlated with at least one trait
sigGeneInfo = subset(geneInfo, paste0("ME", geneInfo$moduleColor) %in% rownames(moduleTraitPvalue[sigRows,]))

```


```{r}
#fix(geneInfo)
```

```{r}
get_top_n = function(mod_color, df, max_n){
  temp = subset(df, df$moduleColor == mod_color)
  column = which(colnames(temp) == paste0("MM.", mod_color))
  temp = temp[order(temp[,column]),]
  subset(temp, abs(temp[,column]) >= max_n)
}
```

```{r}
write.csv(geneInfo, paste0("..\\data_analysis\\", "WGCNA_output\\", params$genome, params$regex, "_module_info.csv"))


for(module in unique(sigGeneInfo$moduleColor)){
  write.csv(subset(sigGeneInfo, sigGeneInfo$moduleColor == module), paste0("..\\data_analysis\\", "WGCNA_output\\", params$genome, params$regex, "_sig_module_", module, "_info.csv"))
  
  write.csv(get_top_n(module, geneInfo, 0.95), paste0("..\\data_analysis\\", "WGCNA_output\\", params$genome, params$regex, "_sig_module_", module, "_top_n_info.csv"))
}

```



```{r}
tMEs = as.data.frame(t(MEs))

tMEs$reg = apply(tMEs, MARGIN = 1, FUN = function(x){ if(median(x) > 0){return("up")}else{return("down")}})

tMEs$module = sapply(strsplit(rownames(tMEs), "ME", fixed = TRUE), function(x) x[2])

tMEs = pivot_longer(tMEs, cols = starts_with(params$regex), names_to = "sample", values_to = "ME")
tMEs = merge(tMEs, data.frame(sample = rownames(datTraits), treatment = c(rep("control", times = sum(datTraits$Control)), rep("relaxation", times = sum(datTraits$Relaxation)), rep("upwelling", times = sum(datTraits$Upwelling)))))
tMEs = merge(tMEs, cbind(datTraits[-c(1:3)], sample = rownames(datTraits)))
tMEs
```


```{r}
#get modules with significant associations with categorical modules
sig_cat_names = mapply(sig_categorical, FUN = function(x){return(x[1])})
#convert index to module names from moduleTraitCor so they are in the same order as the text matrix we pulled the significance info from
sig_cat_names = rownames(moduleTraitCor)[sig_cat_names]
#remove the ME from the start and remove duplicates
sig_cat_names = unique(sapply(strsplit(sig_cat_names, "ME", fixed = TRUE), function(x) x[2]))

#select only modules that have significant associations with categorical variables
sig_cat_tMEs = subset(tMEs, tMEs$module %in% sig_cat_names)



ggplot(sig_cat_tMEs, aes(x = module, y = ME, fill = reg)) + geom_boxplot() + theme(axis.text.x = element_text(angle=45, vjust=1, hjust=1, size = 10)) + ggtitle("Significant module eigengene distributions") + ylab("Module eigengene expression") + geom_hline(yintercept = 0, color = "red") + scale_fill_manual(values = c("up" = "#FF7F7F", "down" = "lightblue"))

ggsave(paste0("..\\data_analysis\\images\\WGCNA\\", params$genome, params$regex, "_sig_ME_dist_boxplot.png"))

ggplot(sig_cat_tMEs, aes(x = as.factor(treatment), y = ME)) + geom_boxplot() + facet_wrap(~module) + geom_hline(yintercept = 0, color = "red") + xlab("treatment") + ggtitle("Significant module eigengene expression by treatment")

ggsave(paste0("..\\data_analysis\\images\\WGCNA\\", params$genome, params$regex, "_sig_ME_treatment_boxplot.png"))
```

```{r}
graph_sig_cont = function(i, sig_continuous, names_list, traits_list, tMEs){
  #get modules with significant associations with categorical modules
  sig_cont_names = unlist(compact(mapply(sig_continuous, FUN = function(x){if(x[2] == i){return(x[1])}})))
  #convert index to module names
  sig_cont_names = names_list[sig_cont_names]
  #remove the ME from the start and remove duplicates
  sig_cont_names = unique(sapply(strsplit(sig_cont_names, "ME", fixed = TRUE), function(x) x[2]))
  
  
  sig_cont_tMEs = subset(tMEs, tMEs$module %in% sig_cont_names)
  
  if(length(sig_cont_tMEs) == 1){
    ggplot(sig_cont_tMEs, aes(x = .data[[traits_list[i]]], y = ME)) + geom_point()
  }else if(length(sig_cont_tMEs > 1)){
    ggplot(sig_cont_tMEs, aes(x = .data[[traits_list[i]]], y = ME)) + geom_point() + facet_wrap(~module)
  }
}
```


```{r}
plot_list = compact(lapply(c(4:length(colnames(moduleTraitCor))), graph_sig_cont, sig_continuous, rownames(moduleTraitCor), colnames(moduleTraitCor), tMEs))

arrange = ggarrange(plots = plot_list)

ggsave(paste0("..\\data_analysis\\images\\WGCNA\\", params$genome, params$regex, "_sig_ME_response_scatterplot.png"), plot = arrange)
```


#Network visualization using WGCNA functions
```{r}
#this chunk may take a while
#dissTOM = 1-TOMsimilarityFromExpr(datExpr, power = spower)
#plotTOM = dissTom^7
#diag(plotTOM) = NA
#sizeGrWindow(9,9)
#TOMplot(plotTOM, geneTree, moduleColors, main = "Network heatmap plot, all genes")
```

```{r}
#visualizing the network of eigengenes

#recalculate module eigengenes
MEs = moduleEigengenes(datExpr, moduleColors)$eigengenes

#Isolate wanted physiological traits from the trait dataframe
#weight = as.data.frame(datTraits$weight_g)
#names(weight) = "weight"

#add the weight to existing module eigengenes
MET = orderMEs(cbind(MEs, weight))
sizeGrWindow(5, 7.5)
par(cex = 0.9)
plotEigengeneNetworks(MET, "", marDendro = c(0,4,1,2), marHeatmap = c(3,4,1,2), cex.lab = 0.8, xLabelsAngle = 90)
#plot the dendrogram
sizeGrWindow(6,6)
par(cex = 1.0)
plotEigengeneNetworks(MET, "Eigengene dendrogram", marDendro = c(0,4,2,0), plotHeatmaps = FALSE)
```


```{r}
#plot the heatmap matrix
par(cex = 1.0)
plotEigengeneNetworks(MET, "Eigengene adjancency heatmap", marHeatmap = c(3,4,2,2), plotDendrograms = FALSE, xLabelsAngle = 90)
```

```{r}
#visualizing the network of significant eigengenes

#recalculate module eigengenes
MEs = moduleEigengenes(datExpr, moduleColors)$eigengenes

#select only significant modules
#MEs$module = sapply(strsplit(names(MEs), "ME", fixed = TRUE), function(x) x[2])
#MEs
sig_MEs = select(MEs, contains(unique(sigGeneInfo$moduleColor)))
#sig_MEs = subset(sig_MEs, select = -c(module))

#Isolate wanted physiological traits from the trait dataframe
#weight = as.data.frame(datTraits$weight_g)
#names(weight) = "weight"

#add the weight to existing module eigengenes
MET = orderMEs(cbind(sig_MEs, weight))
sizeGrWindow(5, 7.5)
par(cex = 0.9)
plotEigengeneNetworks(MET, "", marDendro = c(0,4,1,2), marHeatmap = c(3,4,1,2), cex.lab = 0.8, xLabelsAngle = 90, colorLabels = FALSE)
#plot the dendrogram
sizeGrWindow(6,6)

png(paste0("..\\data_analysis\\images\\WGCNA\\", params$genome, params$regex, "_sig_ME_trait_dendrogram.png"), width = 950, height = 600)

par(cex = 1.0)
plotEigengeneNetworks(MET, "Eigengene dendrogram", marDendro = c(0,4,2,0), plotHeatmaps = FALSE)

dev.off()
```


```{r}
#plot the heatmap matrix
png(paste0("..\\data_analysis\\images\\WGCNA\\", params$genome, params$regex, "_sig_ME_trait_heatmap.png"), width = 950, height = 600)
par(cex = 1.0)
plotEigengeneNetworks(MET, "Eigengene correlation heatmap", marHeatmap = c(13,14,2,2), plotDendrograms = FALSE, xLabelsAngle = 90, colorLabels = FALSE, plotAdjacency = FALSE)

dev.off()
```









